Replace Db Drop impl which shuts down purge task, with explicit shutdown signal

Hodkinson opened the pull request:
Resolves #76

Darksonn commented on src/server.rs:
```diff
@@ -205,6 +206,9 @@ pub async fn run(listener: TcpListener, shutdown: impl Future) -> crate::Result<
     // the `mpsc` channel will close and `recv()` will return `None`.
     let _ = shutdown_complete_rx.recv().await;
 
+    // Signal the 'Db' instance to shutdown the task that purges expired keys
+    db.shutdown_purge_task();
+
     Ok(())
```
I was just about to say that since this returns a `Result`, then the method could return without hitting the `shutdown_purge_task` call, however looking more closely it doesn't seem like the method can ever fail. It probably shouldn't be returning an `Ok(())` in the first place.

Anyway, there's still a situation where we might never hit the `shutdown_purge_task` call, namely the situation where the code panics. A way around this is to wrap `db` in another struct, and to call the method from the destructor of that new struct.

Hodkinson replied:
Makes sense, I removed the `OK(())` from this method. Also added a struct which wraps the `Db`. This way the shutdown is more nicely encapsulated in the db module. I'm not completely happy with the name `DbHolder` though.

Darksonn replied:
How about the name `DbDropGuard`?

Hodkinson replied:
Yes I like that, will change it

Hodkinson commented:
Makes sense, I removed the `OK(())` from this method. Also added a struct which wraps the `Db`. This way the shutdown is more nicely encapsulated in the db module. I'm not completely happy with the name `DbHolder` though.

Darksonn commented on src/db.rs:
```diff
@@ -6,6 +6,16 @@ use std::collections::{BTreeMap, HashMap};
+/// this struct is dropped.
+#[derive(Debug)]
+pub(crate) struct DbHolder {
+    /// The `Db` instance that will be shutdown when this `DbHolder` struct
+    /// is dropped.
```
Shutdown without the space is a noun.
```suggestion
    /// The `Db` instance that will be shut down when this `DbHolder` struct
    /// is dropped.
```

Darksonn commented on src/db.rs:
```diff
@@ -93,6 +103,27 @@ struct Entry {
 }
 
+impl DbHolder {
+    /// Create a new `DbHolder`, wrapping a `Db` instance. When this is dropped
+    /// the `Db`'s purge task will be shutdown.
```
```suggestion
    /// Create a new `DbHolder`, wrapping a `Db` instance. When this is dropped
    /// the `Db`'s purge task will be shut down.
```

Hodkinson replied:
Good spot, all these fixed now.

Darksonn reviewed (approved):
This seems correct!
